kufar/books/main.go

```diff
@@ -358,25 +369,31 @@ func ReadBook(d Database) func(w http.ResponseWriter, r *http.Request) {
 			return
 		}
 
-		row := d.db.QueryRow("select id, name, author, date, userid from books where id = $1", idInt)
+		c := NewMap()
 
-		bk := Book{}
-		err = row.Scan(&bk.ID, &bk.Name, &bk.Author, &bk.Date, &bk.UserID)
-		switch {
-		case err == sql.ErrNoRows:
-			http.NotFound(w, r)
-			return
-		case err != nil:
-			http.Error(w, http.StatusText(http.StatusInternalServerError)+": error scanning db", http.StatusInternalServerError)
-			return
-		}
+		if _, ok := c.get(idInt); !ok {
+			row := d.db.QueryRow("select id, name, author, date, userid from books where id = $1", idInt)
 
-		resp := convertToResponse(bk)
-		w.Header().Set("Content-Type", "application/json")
-		if err := json.NewEncoder(w).Encode(resp); err != nil {
-			http.Error(w, http.StatusText(http.StatusBadRequest)+": Error marshalling json", http.StatusBadRequest)
-			return
+			bk := Book{}
+			err = row.Scan(&bk.ID, &bk.Name, &bk.Author, &bk.Date, &bk.UserID)
+			switch {
+			case err == sql.ErrNoRows:
+				http.NotFound(w, r)
+				return
+			case err != nil:
+				http.Error(w, http.StatusText(http.StatusInternalServerError)+": error scanning db", http.StatusInternalServerError)
+				return
+			}
+			c.set(idInt, bk)
 		}
+
+			resp := convertToResponse(c.Items[idInt])
```

Comment:
All working with cache completely right except this. You need to use your `get` method because it's concurrent safe.
You will have a race here